src/main/groovy/edu/oregonstate/mist/locations/db/CachedFacilDAO.groovy

```diff
@@ -0,0 +1,73 @@
+package edu.oregonstate.mist.locations.db
+
+import com.fasterxml.jackson.core.type.TypeReference
+import com.fasterxml.jackson.databind.ObjectMapper
+import edu.oregonstate.mist.locations.Cache
+import edu.oregonstate.mist.locations.core.FacilLocation
+import groovy.transform.CompileStatic
+import org.skife.jdbi.v2.DBI
+import org.skife.jdbi.v2.exceptions.DBIException
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+import java.sql.SQLException
+
+/**
+ * CachedFacilDAO is a wrapper around FacilDAO that adds caching.
+ */
+@CompileStatic
```

Comment:
I see you use this a lot. Any particular reason?

Reply:
Allow me to respond in the form of a mini-essay ;)

`@CompileStatic` and `@TypeChecked` enable type checking for groovy. The longer i've been programming, the more i've come to appreciate type checking. A program that is type-correct is more likely to be correct. If i mistype a method name or something, the type checker will catch that right away, whereas without a type checker, there won't be any error until that code path happens to be executed. If i'm lucky, the error is on a common path that i hit right away when running the code (or is covered by a unit test). But it could also be the case that the error is on a little-used code path -- some error condition that hardly ever occurs. In that case, the error could lie latent, ready to explode at the worst possible moment.

In the first case, type checking shortens the feedback cycle during development, since the type checker can catch potential errors quicker than running the code. In the second case, type checking can catch errors that might be completely missed. It's true that the type checker can sometimes be overly cautious, catching code that is actually correct but which the type system cannot adequately describe, but i find that most of the time it is sufficient. Sometimes you want the full power of groovy's dynamic typing, and in those few cases `@TypeChecked` can be left off.

Regarding `@TypeChecked` versus `@CompileStatic`, they both perform type checking but `@TypeChecked` compiles down to the same dynamic code as normal groovy, whereas `@CompileStatic` compiles down to something closer to the equivalent java code, which is supposed to be faster. I think i started out mainly using `@TypeChecked`, but these days i tend to prefer `@CompileStatic`—as long as the program is type-correct, might as well get a little speed boost, right? They have slightly different semantics but it usually doesn't make a difference.